### PR TITLE
Support for OpenSearch

### PR DESCRIPTION
**Pull request by todvora**
## Description
This PR adds full support for OpenSearch in Graylog. Currently there logic and handling is similar to ES7, so the majority of the code changes solves dependency injections and search-server detection. 

## Motivation and Context
/jenkins-pr-deps Graylog2/graylog-plugin-enterprise#2819

## How Has This Been Tested?
Full unit and integration tests run, locally manually tried to start and run the graylog server against opensearch instance. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by janheise** [comment]
docker-compose.yml used for my tests:
`version: "3"
 
services:
  mongodb:
    image: "mongo:4.2"
    ports:
      - "27017:27017"
    volumes:
      - "mongoconfigdb:/data/configdb"
      - "mongodb:/data/db"
 
  opensearch:
    image: "opensearchproject/opensearch:1.1.0"
    environment:
      - plugins.security.disabled=true
      - discovery.type=single-node
      - logger.deprecation.level=warn
      - action.auto_create_index=false
      - "OPENSEARCH_JAVA_OPTS=-Xms512m -Xmx512m"
    ports:
      - "9200:9200"
    volumes:
      - "opensearch:/usr/share/opensearch/data"
 
volumes:
  mongoconfigdb:
  mongodb:
  opensearch:
`

**Review by janheise — APPROVED** [review]
LGTM!

**Comment by todvora** [comment]
Rebased and ready for a review from @dennisoelkers :pray:

File: graylog2-server/src/main/java/org/graylog2/configuration/validators/ElasticsearchVersionValidator.java [hunk 1]
```diff
@@ -20,19 +20,21 @@
...
             Version.from(6, 0, 0),
             Version.from(7, 0, 0)
     );
 
     @Override
-    public void validate(String name, Version value) throws ValidationException {
-        if (!SUPPORTED_ES_VERSIONS.contains(value)) {
-            throw new ValidationException("Invalid Elasticsearch version specified in " + name + ": " + value
+    public void validate(String name, SearchVersion value) throws ValidationException {
+        if (!SUPPORTED_ES_VERSIONS.contains(value.version())) {
```
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/configuration/validators/ElasticsearchVersionValidator.java:36]: We could also make `SUPPORTED_ES_VERSIONS` a `List<SearchVersion>` in order to keep the product<->version relationships. Then, at this line we could write the condition as 
```
if (SUPPORTED_ES_VERSIONS.stream().noneMatch(supportedVersion -> supportedVersion.satisfies(value.distribution(), value.version()))) {
```

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
Looks good, just one minor remark.

**Review by dennisoelkers — APPROVED** [review]
Looks good! Let's get this moving!

**Comment by todvora** [comment]
Thank you @dennisoelkers!

**Comment by bendem** [comment]
It doesn't look like this is available in v4. Can this be backported to at least the maintained versions?

**Pull request merged**